runtime/src/accounts_db.rs

```diff
@@ -3396,30 +3411,90 @@ impl AccountsDb {
             .report("external_purge_slots_stats", Some(1000));
     }
 
-    // TODO: This is currently:
-    // 1. Unsafe with scan because it can remove a slot in the middle
-    // of a scan.
-    // 2. Doesn't handle cache flushes that happen during the slot deletion (see comment below).
-    pub fn remove_unrooted_slot(&self, remove_slot: Slot) {
-        if self.accounts_index.is_root(remove_slot) {
-            panic!("Trying to remove accounts for rooted slot {}", remove_slot);
-        }
-
-        // TODO: Handle if the slot was flushed to storage while we were removing the cached
-        // slot above, i.e. it's possible the storage contains partial version of the current
-        // slot. One way to handle this is to augment slots to contain a "version", That way,
-        // 1) We clean older versions via the natural clean() pipeline
-        // without having to call this function out of band.
-        // 2) This deletion doesn't have to block on scan
-        // Reads will then always read the latest version of a slot. Scans will also know
-        // which version their parents because banks will also be augmented with this version,
-        // which handles cases where a deletion of one version happens in the middle of the scan.
+    // TODO: This is currently unsafe with scan because it can remove a slot in the middle
+    /// Remove the set of slots in `remove_slots` from both the cache and storage. This requires
+    /// we know the contents of the slot are either:
+    ///
+    /// 1) Completely in the cache
+    /// 2) Have been completely flushed from the cache
+    ///
+    /// in order to guarantee that when this function returns, the contents of the slot have
+    /// been completely and not only partially removed. Thus synchronization with `flush_slot_cache()`
+    /// through `self.remove_unrooted_slots_synchronization` is necessary.
+    pub fn remove_unrooted_slots(&self, remove_slots: &[Slot]) {
+        let rooted_slots = self
+            .accounts_index
+            .get_rooted_from_list(remove_slots.iter());
+        if !rooted_slots.is_empty() {
+            panic!(
+                "Trying to remove accounts for rooted slots {:?}",
+                rooted_slots
+            );
+        }
+
+        let RemoveUnrootedSlotsSynchronization {
+            slots_under_contention,
+            signal,
+        } = &self.remove_unrooted_slots_synchronization;
+
+        {
+            // Slots that are currently being flushed by flush_slot_cache()
+            let mut contended_slots = slots_under_contention.lock().unwrap();
+
+            // Slots that are currently being flushed by flush_slot_cache() AND
+            // we want to remove in this function
+            let mut contended_cache_flush_slots: Vec<Slot> = remove_slots
+                .iter()
+                .filter(|remove_slot| {
+                    let is_cache_flushing_slot = contended_slots.contains(remove_slot);
```

Comment:
nits: rename to `is_being_flushed` to be consistent with https://github.com/solana-labs/solana/pull/17269/files#r641874795 (`let is_being_flushed = contended_slots.contains(flushing_slot);`)

imo, semantically same expression should result in same name unless there is good reason (like state transition between the occurences). In this case, I think its usage is same for both occurrences in that it's used to reserve slot with `contended_slots` for removal.